File: LRUCache/LRUCache.ts
class ListNode {
  key: number | null;
  val: any;
  prev: ListNode | null;
  next: ListNode | null;

  constructor(key: number | null = null, val: any = null) {
    this.key = key;
    this.val = val;
    this.prev = null;
    this.next = null;
  }
}

export class LRUCache {
  private capacity: number;
  private cache: Map<number, ListNode>;
  private left: ListNode;
  private right: ListNode;

  constructor(capacity: number) {
    this.capacity = capacity;
    this.cache = new Map<number, ListNode>();
    this.left = new ListNode();
    this.right = new ListNode();

    this.left.next = this.right;
    this.right.prev = this.left;
  }

  private remove(node?: ListNode | null): void {
    if (!node) return;

    const prev = node.prev!;
    const next = node.next!;
    prev.next = next;
    next.prev = prev;
  }

  private insert(node?: ListNode | null): void {
    if (!node) return;

    const prev = this.right.prev!;
    const next = this.right;
    prev.next = node;
    next.prev = node;
    node.prev = prev;
    node.next = next;
  }

  get(key: number): any {
    if (this.cache.has(key)) {
      const node = this.cache.get(key);
      this.remove(node);
      this.insert(node);
      return node?.val;
    }
    return -1;
  }

  put(key: number, value: number): void {
    if (this.cache.has(key)) {
      this.remove(this.cache.get(key));
    }
    const newNode = new ListNode(key, value);
    this.cache.set(key, newNode);
    this.insert(newNode);

    if (this.cache.size > this.capacity) {
      const lru = this.left.next;
      this.remove(lru);
      if (lru && lru.key) {
        this.cache.delete(lru.key);
      }
    }
  }
}

File: LRUCache/LRUCache.py
class Node:
    def __init__(self, key=None, val=None):
        self.key, self.val = key, val
        self.prev = self.next = None

class LRUCache:
    
    def remove(self, node):
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev
    
    def insert(self, node):
        prev, nxt = self.right.prev, self.right
        prev.next = nxt.prev = node
        node.prev, node.next = prev, nxt

    def __init__(self, capacity: int):
        self.cp = capacity
        self.cache = {}
        self.left = self.right = Node()
        self.left.next, self.right.prev = self.right, self.left

    def get(self, key: int) -> int:
        if key in self.cache:
            self.remove(self.cache[key])
            self.insert(self.cache[key])
            return self.cache[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])
        self.cache[key] = Node(key, value)
        self.insert(self.cache[key])
        
        if len(self.cache) > self.cp:
            lru = self.left.next
            self.remove(lru)
            del self.cache[lru.key]